### PR TITLE
Added CollectRetryErrorMessagesModule, improved EmailModule

### PR DESCRIPTION
**Pull request by Didddy**
- Added CollectErrorMessagesModule
- Improve handling of invalid string to bool parsing in EmailModule

### CODE REVIEW

**Comment by Didddy** [comment]
@mwergkarner Pls change descriptions where necessary, thx!

File: AutomationHelpers/src/Modules/CollectErrorMessagesModule.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        /// <summary>
+        /// Recursively collects all error messages from given test container
+        /// </summary>
+        /// <param name="container"></param>
+        private static void GetErrorMessages(ITestContainerActivity container)
+        {
+            foreach (var containerChild in container.Children)
+            {
+                // Check if TestContainerActivity
+                var testContainerActivity = containerChild as TestContainerActivity;
```
[the remark this answers is not in the document]
Comment (Didddy) [reply AutomationHelpers/src/Modules/CollectErrorMessagesModule.cs]: Good point! Will adapt...

File: AutomationHelpers/src/Modules/CollectErrorMessagesModule.cs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+                if (testContainerActivity != null)
+                {
+                    GetErrorMessages(testContainerActivity);
+                }
+
+                // Check if TestModuleActivity
+                var testModuleActivity = containerChild as TestModuleActivity;
```
[the remark this answers is not in the document]
Comment (Didddy) [reply AutomationHelpers/src/Modules/CollectErrorMessagesModule.cs]: Good point! Will adapt...

File: AutomationHelpers/src/Modules/CollectErrorMessagesModule.cs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+        private static void AddErrorMessages(
+            string testModuleActivityFullDisplayName,
+            string containerFullDisplayName,
+            IEnumerable<IReportItem> items)
+        {
+            foreach (var testModuleChild in items)
+            {
+                // Check if testModuleChild is ReportItem
+                var reportItem = testModuleChild as ReportItem;
```
[the remark this answers is not in the document]
Comment (Didddy) [reply AutomationHelpers/src/Modules/CollectErrorMessagesModule.cs]: IReportItem does not expose Level and Message, so this has to be used.

File: AutomationHelpers/src/Modules/EmailModule.cs [hunk 4]
```diff
@@ -132,10 +132,23 @@ private void OnTestSuiteCompletedSendResult(object sender, EventArgs e)
             var currentTestSuiteStatus = TestReport.CurrentTestSuiteActivity;
             var reportFile = CreateReports();
 
-            if ((bool.Parse(this.SendEmailOnFailure) && currentTestSuiteStatus.Status == ActivityStatus.Failed)
-                || (bool.Parse(this.SendEmailOnSuccess) && currentTestSuiteStatus.Status == ActivityStatus.Success)
-                || (bool.Parse(this.SendZippedReportOnComplete) && !bool.Parse(this.SendEmailOnFailure) && !bool.Parse(this.SendEmailOnSuccess)))
+            var sendEmailOnFailure = true;
```
[the remark this answers is not in the document]
Comment (Didddy) [reply AutomationHelpers/src/Modules/EmailModule.cs]: Agreed! Will remove unnecessary initializations...

File: AutomationHelpers/src/Modules/EmailModule.cs [hunk 5]
```diff
@@ -132,10 +132,23 @@ private void OnTestSuiteCompletedSendResult(object sender, EventArgs e)
...
+            bool.TryParse(this.SendZippedReportOnComplete, out sendZippedReportOnComplete);
+            var sendPdfReportOnComplete = false;
+            bool.TryParse(this.SendPdfReportOnComplete, out sendPdfReportOnComplete);
+
+            if (sendEmailOnFailure && currentTestSuiteStatus.Status == ActivityStatus.Failed
+                || sendEmailOnSuccess && currentTestSuiteStatus.Status == ActivityStatus.Success
+                || sendZippedReportOnComplete && !sendEmailOnFailure && !sendEmailOnSuccess
+                || sendPdfReportOnComplete && !sendEmailOnFailure && !sendEmailOnSuccess)
             {
+                var useSSL = false;
```
[the remark this answers is not in the document]
Comment (Didddy) [reply AutomationHelpers/src/Modules/EmailModule.cs]: Agreed! Will remove unnecessary initializations...

File: AutomationHelpers/src/Modules/EmailModule.cs [hunk 6]
```diff
@@ -144,19 +157,20 @@ private void OnTestSuiteCompletedSendResult(object sender, EventArgs e)
...
+                    useSSL,
                     this.Username,
                     this.Password);
             }
         }
 
         private string CreateCompressedReport()
         {
-            if (bool.Parse(this.SendZippedReportOnComplete))
+            var sendZippedReportOnComplete = false;
```
[the remark this answers is not in the document]
Comment (Didddy) [reply AutomationHelpers/src/Modules/EmailModule.cs]: Agreed! Will remove unnecessary initializations...

File: AutomationHelpers/src/Modules/EmailModule.cs [hunk 7]
```diff
@@ -168,23 +182,24 @@ private string CreateCompressedReport()
...
-        	if (activity.GetType().Name.Equals("TestSuiteActivity"))
-        	    {
-        	    	TestReport.EndTestModule();
-                	Report.End();
-        	    }
-        	    
-        	    TestReport.SaveReport();
+            Activity activity = ActivityStack.Current;
+
+            if (activity.GetType().Name.Equals("TestSuiteActivity"))
```
[the remark this answers is not in the document]
Comment (Didddy) [reply AutomationHelpers/src/Modules/EmailModule.cs]: Totally agreed! Will adapt...

File: AutomationHelpers/src/Modules/EmailModule.cs [hunk 8]
```diff
@@ -168,23 +182,24 @@ private string CreateCompressedReport()
...
+                Report.End();
+            }
+
+            TestReport.SaveReport();
         }
 
         private string CreatePdfReport()
         {
-            if (bool.Parse(this.SendPdfReportOnComplete))
+            var sendPdfReportOnComplete = false;
```
[the remark this answers is not in the document]
Comment (Didddy) [reply AutomationHelpers/src/Modules/EmailModule.cs]: Agreed! Will remove unnecessary initializations...

File: AutomationHelpers/src/Modules/EmailModule.cs [hunk 9]
```diff
@@ -132,10 +132,23 @@ private void OnTestSuiteCompletedSendResult(object sender, EventArgs e)
             var currentTestSuiteStatus = TestReport.CurrentTestSuiteActivity;
             var reportFile = CreateReports();
 
-            if ((bool.Parse(this.SendEmailOnFailure) && currentTestSuiteStatus.Status == ActivityStatus.Failed)
-                || (bool.Parse(this.SendEmailOnSuccess) && currentTestSuiteStatus.Status == ActivityStatus.Success)
-                || (bool.Parse(this.SendZippedReportOnComplete) && !bool.Parse(this.SendEmailOnFailure) && !bool.Parse(this.SendEmailOnSuccess)))
+            var sendEmailOnFailure = true;
+            if (!bool.TryParse(this.SendEmailOnFailure, out sendEmailOnFailure)) sendEmailOnFailure = true;
```
[the remark this answers is not in the document]
Comment (Didddy) [reply AutomationHelpers/src/Modules/EmailModule.cs]: Totally agreed! Will adapt... I think changing this should be save and not considered as a breaking change?!

[the remark this answers is not in the document]
Comment (ahoisl) [reply AutomationHelpers/src/Modules/EmailModule.cs]: Not sure, sorry. Would require some testing...

**Comment by ahoisl** [comment]
Looks good from my side now

File: AutomationHelpers/src/Modules/CollectRetryErrorMessagesModule.cs [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+using System;
+using System.Collections.Generic;
+using Ranorex.Core.Testing;
+using Ranorex.Core.Reporting;
+
+namespace Ranorex.AutomationHelpers.Modules
+{
+    /// <summary>
+    /// Functionality: Collects all error messages from given test container.
```
Comment (mwergkarner) [inline AutomationHelpers/src/Modules/CollectRetryErrorMessagesModule.cs]: When placed in a test container with a retry count >= 1, collects all error messages even if one of the retries was successful and adds them to the report.

**Pull request merged**